### PR TITLE
Move phpcs from Travis to GitHub Actions

### PR DESCRIPTION
**Pull request by bendavies**
|      Q       |   A
|------------- | -----------
| Type         | improvement
| BC Break     | no
| Fixed issues | 

#### Summary

Moves phpcs from Travis to GitHub Actions.

https://github.com/doctrine/dbal/pull/3978 needs merging first.

### CODE REVIEW

File: .github/workflows/continuous-integration.yml [hunk 1]
```diff
@@ -23,7 +23,7 @@ jobs:
         uses: "actions/cache@v1"
         with:
           path: "~/.composer/cache"
-          key: "composer-${{ hashFiles('composer.json') }}"
+          key: "composer-${{ hashFiles('composer.lock') }}"
```
Comment (greg0ire) [inline .github/workflows/continuous-integration.yml]: :thinking: I will copy that in #3978 I guess. This commit should probably go since `composer update` is still in use here.

File: .github/workflows/continuous-integration.yml [hunk 2]
```diff
@@ -32,7 +32,7 @@ jobs:
           restore-keys: "${{ matrix.php-version }}-composer-locked-"
 
       - name: "Install dependencies with composer"
-        run: "composer update --no-interaction --no-progress --no-suggest"
+        run: "composer install --no-interaction --no-progress --no-suggest"
```
Comment (greg0ire) [inline .github/workflows/continuous-integration.yml:35]: Should be fixed up with previous commit

File: .github/workflows/continuous-integration.yml [hunk 3]
```diff
@@ -36,3 +36,36 @@ jobs:
...
         run: "vendor/bin/psalm"
+
+  coding-standards:
+    name: "Coding Standards"
+    runs-on: "ubuntu-latest"
+
+    strategy:
+      matrix:
+        php-version:
+          - "7.2"
```
Comment (greg0ire) [inline .github/workflows/continuous-integration.yml:47]: Now that you fixed the issue with 7.4 locally, maybe we can use 7.4 here too?

Comment (bendavies) [reply .github/workflows/continuous-integration.yml]: well, the issue seems to be in vendors, but let's try.

File: .github/workflows/continuous-integration.yml (hunk 2 above)
Comment (bendavies) [reply .github/workflows/continuous-integration.yml:35]: done, but will removed once your PR is merged.

File: .github/workflows/continuous-integration.yml (hunk 3 above)
Comment (bendavies) [reply .github/workflows/continuous-integration.yml]: seems fine

**Comment by greg0ire** [comment]
#3978 was merged :)

File: phpcs.xml.dist [hunk 4]
```diff
@@ -2,7 +2,7 @@
 <ruleset>
     <arg name="basepath" value="."/>
     <arg name="extensions" value="php"/>
-    <arg name="parallel" value="80"/>
+    <arg name="parallel" value="8"/>
```
Comment (carusogabriel) [inline phpcs.xml.dist:5]: Why this change?

Comment (bendavies) [reply phpcs.xml.dist:5]: It seems to me that starting 80 processes is a typo copied from somewhere else. Maybe 8 processes was the original intention.

Comment (greg0ire) [reply phpcs.xml.dist:5]: Looks like phpcs itself is run with a similar value: https://github.com/squizlabs/PHP_CodeSniffer/commit/a57331c6c033ef3c947415898b9921a4237289cf

Comment (bendavies) [reply phpcs.xml.dist:5]: Happy to revert. Just seems odd.

Comment (greg0ire) [reply phpcs.xml.dist:5]: Did you see my edit?

Comment (carusogabriel) [reply phpcs.xml.dist:5]: Interesting. I'll research about `parallel` and `phpcs` to see if this change is right because then we'd need to propagate it to the other Doctrine projects. Thanks for taking a look into that :)

Comment (bendavies) [reply phpcs.xml.dist:5]: shall i revert? normally tools like this set the number of parallel process to the number of available cores, or something like that. 80 just seems wild. not sure why PHPCS is using 75 either.

Comment (greg0ire) [reply phpcs.xml.dist:5]: > normally tools like this set the number of parallel process to the number of available cores

That makes sense to me if your process is CPU bound, but if the process spends 90 percent of its time actually being IO bound, it would make sense to me to multiply the number of cores by some factor, so that the scheduler can give the CPU to a process that is no longer waiting for IO. But again, maybe what I'm saying does not make sense. Can you try locally and tell us what figure results in the fastest processing time for you?

Comment (bendavies) [reply phpcs.xml.dist]: I've reverted so we can get this merged but:
```
➜  dbal git:(phpcs-github-actions) ✗ time vendor/bin/phpcs --no-cache --parallel=8
........ 8 / 8 (100%)


Time: 12.85 secs; Memory: 8MB

vendor/bin/phpcs --no-cache --parallel=8  51.49s user 3.37s system 423% cpu 12.945 total
➜  dbal git:(phpcs-github-actions) ✗ time vendor/bin/phpcs --no-cache --parallel=1
............................................................  60 / 456 (13%)
............................................................ 120 / 456 (26%)
............................................................ 180 / 456 (39%)
............................................................ 240 / 456 (53%)
............................................................ 300 / 456 (66%)
............................................................ 360 / 456 (79%)
............................................................ 420 / 456 (92%)
....................................                         456 / 456 (100%)


Time: 28.9 secs; Memory: 34MB

vendor/bin/phpcs --no-cache --parallel=1  28.56s user 0.23s system 99% cpu 28.986 total
➜  dbal git:(phpcs-github-actions) ✗ time vendor/bin/phpcs --no-cache --parallel=8
........ 8 / 8 (100%)


Time: 12.7 secs; Memory: 8MB

vendor/bin/phpcs --no-cache --parallel=8  49.23s user 3.21s system 410% cpu 12.780 total
➜  dbal git:(phpcs-github-actions) ✗ time vendor/bin/phpcs --no-cache --parallel=80
............................................................ 60 / 76 (79%)
................                                             76 / 76 (100%)


Time: 9.2 secs; Memory: 8MB

vendor/bin/phpcs --no-cache --parallel=80  47.74s user 2.92s system 546% cpu 9.264 total
```

Comment (greg0ire) [reply phpcs.xml.dist]: Indeed almost no gain, so this is probably more CPU-bound nowadays.

**Review by greg0ire — APPROVED** [review]

File: .github/workflows/continuous-integration.yml [hunk 5]
```diff
@@ -6,28 +6,65 @@ on:
...
+        with:
+          path: "~/.composer/cache"
+          key: "${{ matrix.php-version }}-composer-locked-${{ hashFiles('composer.lock') }}"
+          restore-keys: "${{ matrix.php-version }}-composer-locked-"
+
+      - name: "Install dependencies with composer"
+        run: "composer install --no-interaction --no-progress --no-suggest"
+
+      - name: "Run squizlabs/php_codesniffer"
+        run: "vendor/bin/phpcs"
```
Comment (morozov) [inline .github/workflows/continuous-integration.yml:70]: Most likely, you'll need to add `--no-colors` here (see https://github.com/mnapoli/pretty/pull/22).

**Review by greg0ire — CHANGES_REQUESTED** [review]
Please add `cs2pr`, I noticed @beberlei did that on ORM (master branch). I'm doing so in a PR to 2.7: https://github.com/doctrine/orm/pull/8116

File: .github/workflows/continuous-integration.yml [hunk 6]
```diff
@@ -6,28 +6,66 @@ on:
...
 
       - name: "Cache dependencies installed with composer"
-        uses: "actions/cache@v1.0.3"
+        uses: "actions/cache@v1"
         with:
           path: "~/.composer/cache"
-          key: "composer-${{ hashFiles('composer.lock') }}"
-          restore-keys: "composer-"
+          key: "${{ matrix.php-version }}-composer-locked-${{ hashFiles('composer.lock') }}"
+          restore-keys: "${{ matrix.php-version }}-composer-locked-"
```
Comment (greg0ire) [inline .github/workflows/continuous-integration.yml:32]:
```suggestion
          key: " php-${{ matrix.php-version }}-composer-locked-${{ hashFiles('composer.lock') }}"
          restore-keys: "php-${{ matrix.php-version }}-composer-locked-"
```

File: .github/workflows/continuous-integration.yml [hunk 7]
```diff
@@ -6,28 +6,66 @@ on:
...
+          coverage: "none"
+          php-version: "${{ matrix.php-version }}"
+          tools: "cs2pr"
+
+      - name: "Cache dependencies installed with composer"
+        uses: "actions/cache@v1"
+        with:
+          path: "~/.composer/cache"
+          key: "${{ matrix.php-version }}-composer-locked-${{ hashFiles('composer.lock') }}"
+          restore-keys: "${{ matrix.php-version }}-composer-locked-"
```
Comment (greg0ire) [inline .github/workflows/continuous-integration.yml:65]:
```suggestion
          key: "php-${{ matrix.php-version }}-composer-locked-${{ hashFiles('composer.lock') }}"
          restore-keys: "php-${{ matrix.php-version }}-composer-locked-"
```

**Review by greg0ire — DISMISSED** [review]
:ok_hand:

**Comment by morozov** [comment]
Is there a way to see how `cs2pr` works in this repository before merging the changes? E.g. push a temporary commit that would violate the coding standard.

File: .github/workflows/continuous-integration.yml [hunk 8]
```diff
@@ -6,28 +6,66 @@ on:
 
 jobs:
   static-analysis-psalm:
-    name: Static Analysis with Psalm
-    runs-on: ubuntu-latest
+    name: "Static Analysis with Psalm"
```
Comment (morozov) [inline .github/workflows/continuous-integration.yml:9]: Are changes like this really needed? I'd rather keep it simple and only add quotes where required (e.g. strings with special characters, numeric strings that must remain strings, etc.).

This file will be maintained by humans who will not see the need to add quotes everywhere.

File: .github/workflows/continuous-integration.yml [hunk 9]
```diff
@@ -6,28 +6,66 @@ on:
...
+          php-version: "${{ matrix.php-version }}"
 
       - name: "Cache dependencies installed with composer"
-        uses: "actions/cache@v1.0.3"
+        uses: "actions/cache@v1"
         with:
           path: "~/.composer/cache"
-          key: "composer-${{ hashFiles('composer.lock') }}"
-          restore-keys: "composer-"
+          key: "php-${{ matrix.php-version }}-composer-locked-${{ hashFiles('composer.lock') }}"
```
Comment (morozov) [inline .github/workflows/continuous-integration.yml:31]: This looks like an unnecessary complication. What's the scenario where we're going to run the code style check using different PHP versions?

File: .github/workflows/continuous-integration.yml [hunk 10]
```diff
@@ -6,28 +6,66 @@ on:
...
-      - name: Psalm
+      - name: "Psalm"
         run: "vendor/bin/psalm"
+
+  coding-standards:
+    name: "Coding Standards"
+    runs-on: "ubuntu-latest"
+
+    strategy:
+      matrix:
```
Comment (morozov) [inline .github/workflows/continuous-integration.yml:45]: I believe the whole point of static analysis is that it's static and doesn't depend on the runtime environment. Having a matrix here looks like overkill to me.

File: .github/workflows/continuous-integration.yml (hunk 8 above)
Comment (bendavies) [reply .github/workflows/continuous-integration.yml:9]: It is trying to maintain consistency across repos
https://github.com/doctrine/orm/pull/8116/files

File: .github/workflows/continuous-integration.yml (hunk 10 above)
Comment (bendavies) [reply .github/workflows/continuous-integration.yml:45]: It is trying to maintain consistency across repos
https://github.com/doctrine/orm/pull/8116/files

File: .github/workflows/continuous-integration.yml (hunk 8 above)
Comment (morozov) [reply .github/workflows/continuous-integration.yml:9]: So what's the reason why we want to have all strings quoted in all repositories?

File: .github/workflows/continuous-integration.yml (hunk 9 above)
Comment (morozov) [reply .github/workflows/continuous-integration.yml:31]: Could you point me to the thread, please? I could find the discussion regarding the build matrix.

File: .github/workflows/continuous-integration.yml (hunk 8 above)
Comment (bendavies) [reply .github/workflows/continuous-integration.yml:9]: i suspect inspiration from https://github.com/ergebnis/php-library-template/blob/master/.github/workflows/integrate.yaml

File: .github/workflows/continuous-integration.yml (hunk 9 above)
Comment (bendavies) [reply .github/workflows/continuous-integration.yml:31]: https://github.com/doctrine/orm/pull/8116#discussion_r418949951

File: .github/workflows/continuous-integration.yml (hunk 8 above)
Comment (bendavies) [reply .github/workflows/continuous-integration.yml:9]: it's just coding style, and that repo enforces it:
https://github.com/ergebnis/php-library-template/blob/master/.yamllint.yaml

I personally don't like everything quoted, as you point out it's not necessary all the time. but being consistent is the main thing i guess.

cc @localheinz

File: .github/workflows/continuous-integration.yml (hunk 9 above)
Comment (bendavies) [reply .github/workflows/continuous-integration.yml:31]: sorry, let me elaborate, because that link doesn't help you.

I think, even if the matrix has a single php version, it enables us to use the same `key` everywhere, which helps cache hits. rather than maintaining 
`key: "php-${{ matrix.php-version }}-composer-locked-${{ hashFiles('composer.lock') }}"` where a matrix is used and
`key: "php-7.4-composer-locked-${{ hashFiles('composer.lock') }}"`
where a matrix is not used.

File: .github/workflows/continuous-integration.yml (hunk 8 above)
Comment (localheinz) [reply .github/workflows/continuous-integration.yml:9]: @morozov 

I am using [`ibiqlik/action-yamllint`](https://github.com/ibiqlik/action-yamllint) to lint YAML files using [`yamllint`](https://github.com/adrienverge/yamllint), and for consistency, I quote YAML strings.

As @bendavies pointed out, it's a matter of taste and consistency. I have to admit that I dislike having to escape a bunch of characters when quoting.

File: .github/workflows/continuous-integration.yml (hunk 9 above)
Comment (morozov) [reply .github/workflows/continuous-integration.yml:31]: So, by “everywhere” you mean that if named consistently, the same cache entry can be shared between multiple actions? Makes sense in this case.

**Comment by morozov** [comment]
Looks good. Should we test https://github.com/doctrine/dbal/pull/3979#issuecomment-624343839 and squash everything before the merge?

File: .github/workflows/continuous-integration.yml (hunk 9 above)
Comment (bendavies) [reply .github/workflows/continuous-integration.yml:31]: Multiple action in the same repo yes. Thanks for phrasing it better!

**Comment by bendavies** [comment]
Sure. I can do it later, or feel free to push to my branch if you are available now.

**Comment by greg0ire** [comment]
Screenshot_2020-05-06 Move phpcs from Travis to GitHub Actions by bendavies · Pull Request #3979 · doctrine dbal

**Review by greg0ire — APPROVED** [review]

**Comment by greg0ire** [comment]
After merging up all the way until master, let's add this check as well as Psalm as required, shall we?

**Comment by greg0ire** [comment]
Thanks a lot @bendavies !
